### ISSUE TITLE
Investigate 100% CPU usage at start-up on single-core Windows machines

### ISSUE THREAD

**Issue by leafo** [itchio/itch#1360]
Minimally, we should probably stop it if the window isn't visible

https://twitter.com/mwegner/status/889195402880208896

**Comment by fasterthanlime** [comment]
Single-core windows computer?

If that's the case - we've observed near-100% CPU usage around the time the app is starting up, and I've never been able to track it down.

If not, I have no idea either. I've tried to reproduce the issue and wasn't able to. (For me it just falls down to <1% CPU usage, as it should).

**Comment by mwegner** [comment]
Yeah, it's one of our build VMs, and only has one core available (I'm just using the Itch app to get butler installed/updated automatically).  Here are the at-a-glance details:

![screenshot 2017-07-23 14 47 53](https://user-images.githubusercontent.com/271924/28503150-fcb7f7ea-6fb5-11e7-92bd-d1fc4f7ee897.png)

It's 100% reproducible on my end, so happy to provide a process dump or whatever else might help track this down!  It takes about two minutes of 100% CPU consumption to show any kind of UI, and then continues at 100% while sitting on the login screen, either in foreground or minimized.

This is a VM, but the ESXi host only has two VMs on it, and at the moment both are idle.  It's an overclocked G4560, so pretty decent ~2500 passmark score at 4.2GHz.  Should be pretty speedy!

**Comment by fasterthanlime** [comment]
> I'm just using the Itch app to get butler installed/updated automatically

Oh! As much as I would love to solve the single-core issue, you definitely don't need the app for that:

  * https://itch.io/docs/butler/installing.html
  * https://itch.io/docs/butler/upgrading.html

TL;DR if you want the latest bleeding-edge, just download

  * https://dl.itch.ovh/butler/windows-amd64/head/butler.exe

If you want the latest stable, use https://dl.itch.ovh/butler/windows-amd64/LATEST to know the version number:

![image](https://user-images.githubusercontent.com/7998310/28518639-daafa7e8-7068-11e7-84a5-3398976d4f02.png)

I've opened an issue internally to have stable URLs one can use in CI instead of having to do the "get LATEST, then get the real file) dance, so that'll arrive too.

---

Regarding the original issue: it definitely has nothing to do with CSS animations (I'll change the issue title now). Since a lot of code has changed in the new app, I'm not entirely sure it still happens on master, so I'll play around with that today.

@GranPC and I already tracked this one down for a few days earlier, but there's better tools in the latest Electron versions, so let's hope we'll have more success now.

**Comment by fasterthanlime** [comment]
Investigation results so far:

  * My laptop's SSD is way too tiny to afford running a windows VM
  * I was able to run one anyway
  * It's the node process, not the chromium process, which is doing 99% cpu usage
  * It's not that easily reproducible for me - it was the three first times, and then it dropped to 1%, then after fiddling some more it jumped back up to 99%
  * I've done some profiling in the renderer process and there was a lot of calls to `createSelector` that shouldn't have been there

Now I'm puzzled because the renderer process shouldn't have been showing anything funny, since it's the node process that's hogging all the CPU. Maybe it's making a blocking remote call and the node process has a lot of work to do to reply to it? (Like serialize a lot of data to JSON?)

I'm starting to have ideas, I'm also starting to think that it wouldn't happen with v25, which I'm going to check as soon as I figure out how to free up some more space on my tiny, tiny SSD. 💾

**Comment by mwegner** [comment]
> Oh! As much as I would love to solve the single-core issue, you definitely don't need the app for that:

Getting off topic from the single-core issue, but my understanding is that manually installing doesn't have any auto-update logic (and will turn off butler auto-updating via the Itch app if you do have it installed)?

**Comment by fasterthanlime** [comment]
> Getting off topic from the single-core issue, but my understanding is that manually installing doesn't have any auto-update logic (and will turn off butler auto-updating via the Itch app if you do have it installed)?

butler has its own `upgrade` command, which you need to run manually. It'll tell you when it's out-of-date whenever you run a push.

itch updates dependencies like butler every time it starts up. (later versions of the app might update it while it's running - but right now it doesn't do that).

> and will turn off butler auto-updating via the Itch app if you do have it installed

If you drop a bleeding-edge `butler.exe` (ie. `butler -V` returns `head, no build date`) into `%APPDATA%/itch/bin/`, then the app will not attempt to update it.

Any butler executables you may have anywhere else on the computer aren't touched by the app at all.

Hope that clears things up!

**Comment by fasterthanlime** [comment]
After a lot of investigation work, I'm 99% confident this is a known electron issue that comes from a bug in their node event loop integration (that's how @zcbenz describes it, I don't have the details!)

More info on these here:

  * https://github.com/electron/electron/issues/1833 (the canonical issue for this)
  * https://github.com/electron/electron/issues/7839 (has links to two more issues)

To the best of my knowledge, it affects:

  * Signle-core machines only
  * Windows 7 & Windows Server 2012 only

Marking as "blocked" since it's not an itch bug, and we'll bump to whichever electron version ships a fix as soon as it comes out.

**Comment by fasterthanlime** [comment]
Apparently (!) that issue no longer exists with itch v25. @GranPC tested it on a single-core VM and it worked well.
